packages/core/src/mm/actors/Player.c

```diff
@@ -1644,3 +1738,437 @@ void Player_UseItem(GameState_Play* play, Actor_Player* this, s16 itemId)
+    Vec3f* D_801C0994 = (Vec3f*)0x801C0994;
+    f32* sMeleeWeaponLengths = (f32*)0x801C0D78;
+    PlayerMeleeWeapon (*Player_GetMeleeWeaponHeld)(Actor_Player* this) = (void*)0x80124190;
+
+    if (player->itemAction == PLAYER_CUSTOM_IA_HAMMER) {
+        D_801C0994->x = 2500.0f;
+    } else if (player->itemAction == PLAYER_IA_DEKU_STICK) {
+        D_801C0994->x = player->unk_B0C * 5000.0f;
+    } else {
+        D_801C0994->x = sMeleeWeaponLengths[Player_GetMeleeWeaponHeld(player)];
+    }
+}
+
+static MeleeWeaponDamageInfo megatonHammerDmgInfo = { DMG_GORON_PUNCH, 2, 4, 2, 4 };
+
+void Player_SetMeleeWeaponInfo(Actor_Player* this, PlayerMeleeWeaponAnimation meleeWeaponAnim) {
+    MeleeWeaponDamageInfo* D_8085D09C = OverlayAddr(0x8085D09C);
+    MeleeWeaponDamageInfo* dmgInfo = &D_8085D09C[0];
+    PlayerMeleeWeapon (*Player_GetMeleeWeaponHeld)(Actor_Player* this) = (void*)0x80124190;
+    void (*func_80833728)(Actor_Player* this, s32 index, u32 dmgFlags, s32 damage) = OverlayAddr(0x80833728);
+
+    s32 damage;
+
+    if (this->actor.id == AC_EN_TEST3) {
+        /* Was Kafei originally intended to be able to punch? (Part of the original function)*/
+        meleeWeaponAnim = PLAYER_MWA_GORON_PUNCH_LEFT;
+        this->meleeWeaponAnimation = -1;
+    } else if (this->heldItemAction == PLAYER_CUSTOM_IA_HAMMER) {
+        dmgInfo = &megatonHammerDmgInfo;
+    } else {
+        dmgInfo = &D_8085D09C[(this->transformation == MM_PLAYER_FORM_GORON) ? PLAYER_MELEEWEAPON_NONE
+                                                                          : Player_GetMeleeWeaponHeld(this)];
+    }
+
+    damage =
+        ((meleeWeaponAnim >= PLAYER_MWA_FLIPSLASH_START) && (meleeWeaponAnim <= PLAYER_MWA_ZORA_JUMPKICK_FINISH))
+            ? ((this->transformation == MM_PLAYER_FORM_HUMAN) ? dmgInfo->dmgHumanStrong : dmgInfo->dmgTransformedStrong)
+            : ((this->transformation == MM_PLAYER_FORM_HUMAN) ? dmgInfo->dmgHumanNormal : dmgInfo->dmgTransformedNormal);
+
+    func_80833728(this, 0, dmgInfo->dmgFlags, damage);
+    func_80833728(this, 1, dmgInfo->dmgFlags, damage);
+}
+
+PATCH_FUNC(0x8083375C, Player_SetMeleeWeaponInfo)
+
+static AttackAnimInfo sHammerAttackAnimInfo[] = {
+    /* PLAYER_MWA_HAMMER_FORWARD */
+    //{ &gPlayerAnim_link_hammer_hit, &gPlayerAnim_link_hammer_hit_end, &gPlayerAnim_link_hammer_hit_endR, 3, 10 },
+    { (PlayerAnimationHeader*)0x0400E4E8, (PlayerAnimationHeader*)0x0400E4F0, (PlayerAnimationHeader*)0x0400E4E0, 3, 10 },
+    /* PLAYER_MWA_HAMMER_SIDE */
+    //{ &gPlayerAnim_link_hammer_side_hit, &gPlayerAnim_link_hammer_side_hit_end, &gPlayerAnim_link_hammer_side_hit_endR, 2, 11 },
+    { (PlayerAnimationHeader*)0x0400E4D8, (PlayerAnimationHeader*)0x0400E4D0, (PlayerAnimationHeader*)0x0400E4C8, 2, 11 },
+};
+
+AttackAnimInfo* Player_GetMeleeAttackAnimInfo(void* a0, Actor_Player* player, PlayerMeleeWeaponAnimation meleeWeaponAnim) {
+    asm("sw $a1, 0x0018 ($sp)"); /* keep player in a1 */
```

Comment:
Are we okay with this? Is there no way around this or is it fine as is? I suppose it's not much different then having an asm hook right? Like
```
Player_GetMeleeAttackAnimInfoHook:
  addiu sp, sp, -0x20
  // store ra
  jal Player_GetMeleeAttackAnimInfo
    // store a1
  // load a1
  // load ra
  jr ra
  addiu sp, sp, 0x20
```

Reply:
Though this way does assume that 0x0018 (sp) inside the code of this function is available.

Reply:
an asm hook could be used instead yes, I think it's okay though. I'm kind of attached to it now though so I'm biased. I stepped through this function in a debugger to see how the compiler wrote it before and after making this.

Reply:
Yeah big no for this, very unsafe, could break at pretty much any time. Can't you just wrap the func?